scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -863,12 +863,27 @@
                         "in Datomic without actually having been launched in Mesos"
                         matches)
               (throw e))))
-        (log/info "In" pool-name "pool, launching" (count task-txns) "tasks")
+
         (ratelimit/spend! ratelimit/global-job-launch-rate-limiter ratelimit/global-job-launch-rate-limiter-key (count task-txns))
-        (let [num-offers-matched (->> matches
-                                      (mapcat (comp :id :offer :leases))
-                                      (distinct)
-                                      (count))]
+        (let [offers-matched (->> matches
+                                  (mapcat :leases)
+                                  (map :offer))
+              num-offers-matched (->> offers-matched
+                                      (map (comp :value :id))
+                                      distinct
+                                      count)
+              user->num-jobs (->> matches
+                                  (mapcat :task-metadata-seq)
+                                  (map (comp tools/job-ent->user :job :task-request))
+                                  frequencies)
+              hostnames-matched (->> offers-matched
+                                     (map :hostname)
+                                     distinct)]
+          (log/info "In" pool-name "pool, launching matched tasks"
+                    {:hostnames-matched hostnames-matched
+                     :number-offers-matched num-offers-matched
+                     :number-tasks (count task-txns)
+                     :user->number-jobs user->num-jobs})
```

Comment:
Here's a sample `matches` structure:

```
~~~~~ matches =
({:leases
  ({:offer
    {:offer-match-timer
     #object[com.codahale.metrics.Timer$Context 0x24c79378 "com.codahale.metrics.Timer$Context@24c79378"],
     :hostname "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9",
     :compute-cluster "<snip>",
     :executor-ids [],
     :reject-after-match-attempt true,
     :id {:value "ef21cd12-a79e-48bb-8377-0c820418e62b"},
     :slave-id
     {:value "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9"},
     :framework-id "gke-2",
     :resources
     [{:name "mem", :type :value-scalar, :scalar 937.125}
      {:name "cpus", :type :value-scalar, :scalar 0.7389999999999999}
      {:name "disk", :type :value-scalar, :scalar 0.0}
      {:name "gpus", :type :value-text->scalar, :text->scalar {}}],
     :attributes
     ({:name "compute-cluster-type",
       :type :value-text,
       :text "kubernetes"}
      {:name "beta.kubernetes.io/arch",
       :type :value-text,
       :text "amd64"}
      {:name "beta.kubernetes.io/fluentd-ds-ready",
       :type :value-text,
       :text "true"}
      {:name "beta.kubernetes.io/instance-type",
       :type :value-text,
       :text "g1-small"}
      {:name "beta.kubernetes.io/os", :type :value-text, :text "linux"}
      {:name "cloud.google.com/gke-nodepool",
       :type :value-text,
       :text "cook-pool-k8s-alpha-g1-small"}
      {:name "cloud.google.com/gke-os-distribution",
       :type :value-text,
       :text "cos"}
      {:name "cook-pool", :type :value-text, :text "k8s-alpha"}
      {:name "failure-domain.beta.kubernetes.io/region",
       :type :value-text,
       :text "us-central1"}
      {:name "failure-domain.beta.kubernetes.io/zone",
       :type :value-text,
       :text "us-central1-a"}
      {:name "kubernetes.io/arch", :type :value-text, :text "amd64"}
      {:name "kubernetes.io/hostname",
       :type :value-text,
       :text "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9"}
      {:name "kubernetes.io/os", :type :value-text, :text "linux"}
      {:name "node-type", :type :value-text, :text "g1-small"}
      {:name "test-label", :type :value-text, :text "true"})},
    :time 1599256724147}),
  :tasks
  #{#object[com.netflix.fenzo.TaskAssignmentResult 0x382ad9a8 "TaskAssignmentResult{host=gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9, request=cook.scheduler.scheduler.TaskRequestAdapter@ce72e4e7, taskId='5f52b894-5b63-4dd1-8148-8cebfac38199', hostname='gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9', assignedPorts=[], successful=true, failures=null, constraintFailure=null, fitness=0.7317197515000587}"] #object[com.netflix.fenzo.TaskAssignmentResult 0x1a6da818 "TaskAssignmentResult{host=gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9, request=cook.scheduler.scheduler.TaskRequestAdapter@8b9fe392, taskId='5f52b894-ac9a-4e84-a407-ed38d3d6366f', hostname='gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9', assignedPorts=[], successful=true, failures=null, constraintFailure=null, fitness=0.5772399171666863}"] #object[com.netflix.fenzo.TaskAssignmentResult 0x413d3912 "TaskAssignmentResult{host=gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9, request=cook.scheduler.scheduler.TaskRequestAdapter@e45aff1c, taskId='5f52b894-9917-4012-866d-65d953787ad1', hostname='gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9', assignedPorts=[], successful=true, failures=null, constraintFailure=null, fitness=0.6544798343333725}"]},
  :hostname "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9",
  :task-metadata-seq
  ({:labels {},
    :task-id "5f52b894-9917-4012-866d-65d953787ad1",
    :executor-key :container-command-executor,
    :name "vagrant_job_vagrant_5f52b894-9917-4012-866d-65d953787ad1",
    :command
    {:environment
     {"COOK_INSTANCE_NUM" "0",
      "COOK_INSTANCE_UUID" "5f52b894-9917-4012-866d-65d953787ad1",
      "COOK_JOB_UUID" "20090421-9a67-495e-973e-c88fc939287f",
      "COOK_JOB_CPUS" "0.1",
      "COOK_JOB_MEM_MB" "32.0"},
     :uris [],
     :user "root",
     :value "ls"},
    :hostname "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9",
    :ports-assigned [],
    :container
    {:type "DOCKER",
     :docker
     {:image "gcr.io/google-containers/alpine-with-bash:1.0",
      :parameters #{{:key "user", :value "1000:1000"}},
      :network "HOST",
      :force-pull-image false}},
    :environment
    {"COOK_INSTANCE_NUM" "0",
     "COOK_INSTANCE_UUID" "5f52b894-9917-4012-866d-65d953787ad1",
     "COOK_JOB_UUID" "20090421-9a67-495e-973e-c88fc939287f",
     "COOK_JOB_CPUS" "0.1",
     "COOK_JOB_MEM_MB" "32.0"},
    :task-request
    {:job
     {:job/pool
      {:pool/name "k8s-alpha",
       :pool/purpose "This is a pool for testing purposes",
       :pool/state :pool.state/active,
       :pool/dru-mode :pool.dru-mode/default},
      :job/commit-latch
      {:commit-latch/committed? true,
       :commit-latch/uuid
       #uuid "5f52b859-5e9c-4637-addb-fe562aee1af0"},
      :job/submit-time #inst "2020-09-04T21:57:45.264-00:00",
      :job/custom-executor false,
      :job/application
      {:application/name "cook-scheduler-cli",
       :application/version "3.3.0"},
      :job/container
      {:container/type "DOCKER",
       :container/docker
       {:docker/image "gcr.io/google-containers/alpine-with-bash:1.0",
        :docker/parameters
        #{{:docker.param/key "user", :docker.param/value "1000:1000"}},
        :docker/network "HOST",
        :docker/force-pull-image false}},
      :job/state :job.state/waiting,
      :job/name "vagrant_job",
      :job/command "ls",
      :job/max-retries 1,
      :job/resource
      #{{:resource/type :resource.type/cpus, :resource/amount 0.1}
        {:resource/type :resource.type/mem, :resource/amount 32.0}},
      :job/uuid #uuid "20090421-9a67-495e-973e-c88fc939287f",
      :job/user "vagrant",
      :job/disable-mea-culpa-retries false},
     :resources {:ports 0, :cpus 0.2, :mem 160.0},
     :task-id "5f52b894-9917-4012-866d-65d953787ad1",
     :assigned-resources #<Atom@6221d79a: nil>,
     :guuid->considerable-cotask-ids
     #object[cook.tools$make_guuid__GT_considerable_cotask_ids$fn__5742 0x37e2418b "cook.tools$make_guuid__GT_considerable_cotask_ids$fn__5742@37e2418b"],
     :constraints
     (#object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x2436e0e9 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@2436e0e9"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x21449831 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@21449831"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x5cbfa166 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@5cbfa166"]
      #object[cook.scheduler.constraints$build_launch_max_tasks_constraint$reify__9849 0x304e306c "cook.scheduler.constraints$build_launch_max_tasks_constraint$reify__9849@304e306c"]
      #object[cook.scheduler.constraints$build_max_tasks_per_host_constraint$reify__9852 0x23ecaf24 "cook.scheduler.constraints$build_max_tasks_per_host_constraint$reify__9852@23ecaf24"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x26ec2e78 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@26ec2e78"]),
     :scalar-requests {"cpus" 0.1, "mem" 32.0}},
    :executor :executor/mesos,
    :num-ports 0,
    :resources {:mem 32.0, :cpus 0.1},
    :data [123, 58, 105, 110, 115, 116, 97, 110, 99, 101, 32, 48, 125]}
   {:labels {},
    :task-id "5f52b894-5b63-4dd1-8148-8cebfac38199",
    :executor-key :container-command-executor,
    :name "vagrant_job_vagrant_5f52b894-5b63-4dd1-8148-8cebfac38199",
    :command
    {:environment
     {"COOK_INSTANCE_NUM" "0",
      "COOK_INSTANCE_UUID" "5f52b894-5b63-4dd1-8148-8cebfac38199",
      "COOK_JOB_UUID" "20090421-d2b3-4022-82f5-ba2619df195f",
      "COOK_JOB_CPUS" "0.1",
      "COOK_JOB_MEM_MB" "32.0"},
     :uris [],
     :user "root",
     :value "ls"},
    :hostname "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9",
    :ports-assigned [],
    :container
    {:type "DOCKER",
     :docker
     {:image "gcr.io/google-containers/alpine-with-bash:1.0",
      :parameters #{{:key "user", :value "1000:1000"}},
      :network "HOST",
      :force-pull-image false}},
    :environment
    {"COOK_INSTANCE_NUM" "0",
     "COOK_INSTANCE_UUID" "5f52b894-5b63-4dd1-8148-8cebfac38199",
     "COOK_JOB_UUID" "20090421-d2b3-4022-82f5-ba2619df195f",
     "COOK_JOB_CPUS" "0.1",
     "COOK_JOB_MEM_MB" "32.0"},
    :task-request
    {:job
     {:job/pool
      {:pool/name "k8s-alpha",
       :pool/purpose "This is a pool for testing purposes",
       :pool/state :pool.state/active,
       :pool/dru-mode :pool.dru-mode/default},
      :job/commit-latch
      {:commit-latch/committed? true,
       :commit-latch/uuid
       #uuid "5f52b85a-7bfd-4d73-9dbb-53ea18889a02"},
      :job/submit-time #inst "2020-09-04T21:57:46.086-00:00",
      :job/custom-executor false,
      :job/application
      {:application/name "cook-scheduler-cli",
       :application/version "3.3.0"},
      :job/container
      {:container/type "DOCKER",
       :container/docker
       {:docker/image "gcr.io/google-containers/alpine-with-bash:1.0",
        :docker/parameters
        #{{:docker.param/key "user", :docker.param/value "1000:1000"}},
        :docker/network "HOST",
        :docker/force-pull-image false}},
      :job/state :job.state/waiting,
      :job/name "vagrant_job",
      :job/command "ls",
      :job/max-retries 1,
      :job/resource
      #{{:resource/type :resource.type/cpus, :resource/amount 0.1}
        {:resource/type :resource.type/mem, :resource/amount 32.0}},
      :job/uuid #uuid "20090421-d2b3-4022-82f5-ba2619df195f",
      :job/user "vagrant",
      :job/disable-mea-culpa-retries false},
     :resources {:ports 0, :cpus 0.2, :mem 160.0},
     :task-id "5f52b894-5b63-4dd1-8148-8cebfac38199",
     :assigned-resources #<Atom@7f90f52: nil>,
     :guuid->considerable-cotask-ids
     #object[cook.tools$make_guuid__GT_considerable_cotask_ids$fn__5742 0x37e2418b "cook.tools$make_guuid__GT_considerable_cotask_ids$fn__5742@37e2418b"],
     :constraints
     (#object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x46758291 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@46758291"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x2680f4 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@2680f4"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x47fef166 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@47fef166"]
      #object[cook.scheduler.constraints$build_launch_max_tasks_constraint$reify__9849 0x832e554 "cook.scheduler.constraints$build_launch_max_tasks_constraint$reify__9849@832e554"]
      #object[cook.scheduler.constraints$build_max_tasks_per_host_constraint$reify__9852 0x62398fc9 "cook.scheduler.constraints$build_max_tasks_per_host_constraint$reify__9852@62398fc9"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x562c825a "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@562c825a"]),
     :scalar-requests {"cpus" 0.1, "mem" 32.0}},
    :executor :executor/mesos,
    :num-ports 0,
    :resources {:mem 32.0, :cpus 0.1},
    :data [123, 58, 105, 110, 115, 116, 97, 110, 99, 101, 32, 48, 125]}
   {:labels {},
    :task-id "5f52b894-ac9a-4e84-a407-ed38d3d6366f",
    :executor-key :container-command-executor,
    :name "vagrant_job_vagrant_5f52b894-ac9a-4e84-a407-ed38d3d6366f",
    :command
    {:environment
     {"COOK_INSTANCE_NUM" "0",
      "COOK_INSTANCE_UUID" "5f52b894-ac9a-4e84-a407-ed38d3d6366f",
      "COOK_JOB_UUID" "20090421-e8cc-49fa-85c0-e159fa63b8f1",
      "COOK_JOB_CPUS" "0.1",
      "COOK_JOB_MEM_MB" "32.0"},
     :uris [],
     :user "root",
     :value "ls"},
    :hostname "gke-dpo-test-0714-19-cook-pool-k8s-al-18d55bc3-7qb9",
    :ports-assigned [],
    :container
    {:type "DOCKER",
     :docker
     {:image "gcr.io/google-containers/alpine-with-bash:1.0",
      :parameters #{{:key "user", :value "1000:1000"}},
      :network "HOST",
      :force-pull-image false}},
    :environment
    {"COOK_INSTANCE_NUM" "0",
     "COOK_INSTANCE_UUID" "5f52b894-ac9a-4e84-a407-ed38d3d6366f",
     "COOK_JOB_UUID" "20090421-e8cc-49fa-85c0-e159fa63b8f1",
     "COOK_JOB_CPUS" "0.1",
     "COOK_JOB_MEM_MB" "32.0"},
    :task-request
    {:job
     {:job/pool
      {:pool/name "k8s-alpha",
       :pool/purpose "This is a pool for testing purposes",
       :pool/state :pool.state/active,
       :pool/dru-mode :pool.dru-mode/default},
      :job/commit-latch
      {:commit-latch/committed? true,
       :commit-latch/uuid
       #uuid "5f52b858-79b6-4232-81ef-f6c5464f5266"},
      :job/submit-time #inst "2020-09-04T21:57:44.483-00:00",
      :job/custom-executor false,
      :job/application
      {:application/name "cook-scheduler-cli",
       :application/version "3.3.0"},
      :job/container
      {:container/type "DOCKER",
       :container/docker
       {:docker/image "gcr.io/google-containers/alpine-with-bash:1.0",
        :docker/parameters
        #{{:docker.param/key "user", :docker.param/value "1000:1000"}},
        :docker/network "HOST",
        :docker/force-pull-image false}},
      :job/state :job.state/waiting,
      :job/name "vagrant_job",
      :job/command "ls",
      :job/max-retries 1,
      :job/resource
      #{{:resource/type :resource.type/cpus, :resource/amount 0.1}
        {:resource/type :resource.type/mem, :resource/amount 32.0}},
      :job/uuid #uuid "20090421-e8cc-49fa-85c0-e159fa63b8f1",
      :job/user "vagrant",
      :job/disable-mea-culpa-retries false},
     :resources {:ports 0, :cpus 0.2, :mem 160.0},
     :task-id "5f52b894-ac9a-4e84-a407-ed38d3d6366f",
     :assigned-resources #<Atom@266b53f3: nil>,
     :guuid->considerable-cotask-ids
     #object[cook.tools$make_guuid__GT_considerable_cotask_ids$fn__5742 0x37e2418b "cook.tools$make_guuid__GT_considerable_cotask_ids$fn__5742@37e2418b"],
     :constraints
     (#object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x670e2656 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@670e2656"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x20aeb9cd "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@20aeb9cd"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x26cbb9bc "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@26cbb9bc"]
      #object[cook.scheduler.constraints$build_launch_max_tasks_constraint$reify__9849 0x5a43825d "cook.scheduler.constraints$build_launch_max_tasks_constraint$reify__9849@5a43825d"]
      #object[cook.scheduler.constraints$build_max_tasks_per_host_constraint$reify__9852 0x4ce923d5 "cook.scheduler.constraints$build_max_tasks_per_host_constraint$reify__9852@4ce923d5"]
      #object[cook.scheduler.constraints$fenzoize_job_constraint$reify__9855 0x195c8b54 "cook.scheduler.constraints$fenzoize_job_constraint$reify__9855@195c8b54"]),
     :scalar-requests {"cpus" 0.1, "mem" 32.0}},
    :executor :executor/mesos,
    :num-ports 0,
    :resources {:mem 32.0, :cpus 0.1},
    :data
    [123, 58, 105, 110, 115, 116, 97, 110, 99, 101, 32, 48, 125]})})
```

Reply:
And here's the code I added to dump that. The compute cluster is too huge to be useful.

```
        (println "~~~~~ matches =")
        (clojure.pprint/pprint (map (fn [{:keys [leases] :as match}]
                                      (assoc match
                                        :leases
                                        (map (fn [lease]
                                               (assoc-in lease [:offer :compute-cluster] "<snip>"))
                                             leases)))
                                    matches))
```